redis/credentials.py

```diff
@@ -0,0 +1,50 @@
+from typing import Callable, Optional
+
+
+class CredentialsProvider:
+    def __init__(
+        self,
+        username: str = "",
+        password: str = "",
+        supplier: Optional[Callable] = None,
+        *args,
+        **kwargs,
+    ):
+        """
+        Initialize a new Credentials Provider.
+        :param supplier: a supplier function that returns the username and password.
+                         def supplier(arg1, arg2, ...) -> (username, password)
+                         For examples see examples/connection_examples.ipynb
+        :param args: arguments to pass to the supplier function
+        :param kwargs: keyword arguments to pass to the supplier function
+        """
+        self._username = "" if username is None else username
+        self._password = "" if password is None else password
+        self.supplier = supplier
+        self.args = args
```

Comment:
Args and kwargs apply to a function. Ideally at a class level these items would have specified names that aren't this generic.